Support Elastic Common Schema in OpenTelemetry (#199)
open-telemetry/oteps: text/0000-support-elastic-common-schema-in-opentelemetry.md

# Support Elastic Common Schema in OpenTelemetry

Collaborators: Alolita Sharma (OTEL GC, AWS), Cyrille Le Clerc (Elastic), Daniel Khan (Elastic), Christian Beedgen (Sumo), Jonah Kowall (Logz.io), Tigran N. (Splunk)

## Introduction

This proposal is to add support for the Elastic Common Schema (ECS) in the OpenTelemetry specification and provide full interoperability for ECS in OpenTelemetry component implementations.

> 

Author: ```suggestion
The addition would be done enriching the OpenTelemetry Semantic Conventions with the fields defined in ECS. When overlaps are identified, a merge decision will be made favoring the backward compatibility of the OpenTelemetry Semantic Conventions.
```

## Customer Motivation

Adoption of OTEL logs will accelerate greatly if ECS is leveraged as the common standard, using this basis for normalization. OTEL Logs adoption will be accelerated by this support. For example, ECS can provide the unified structured format for handling vendor generated along with open source logs.

Customers will benefit from turnkey logs integrations that will be fully recognized by OTEL compatible observability products and services.

> OpenTelemetry logging is today mostly structured when instrumentation libraries are used. However, most of the logs which exist today are generated by software, hardware, and cloud services which the user cannot control. OpenTelemetry provides a limited set of "reference integrations" to  structure logs: primarily the [OpenTelemetry Collector Kubernetes Events Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8seventsreceiver) and an example of a regexp based parsing of Tomcat access logs with OpenTelemetry Collector File Receiver ([here](https://github.com/open-telemetry/opentelemetry-log-collection/blob/30807b96b2f0771e7d11452ebf98fe5e211ed6d7/examples/tomcat/config.yaml#L20)).

Reviewer: It is not clear to me how this proposal to add ECS fields in the OTel semconv would address the challenge described here. Can you expand on this?

Reviewer 2: The purpose of ECS is to provide correlation between different log formats. For example, if you are looking for a log about a dropped firewall event across multiple firewalls. The log message is represented in a similar form in ECS versus the way that Linux, Cisco, and Palo Alto might format the firewall log. This allows us to correlate data with a single query versus trying to build complex queries (and parsing rules) on the backend to correlate. This purpose of this is not for otel instrumented code, but for non-instrumented code.

Reviewer: Thank you for the explanation. The expectation is not that such vendors would pick up ECS or OTel semconv as their logging format in any foreseeable future yet (and certainly not for already deployed appliances) but that we would be able to define transformation/mapping rules that are then implemented in the respective OTel collector receivers or processors, right?
Do such mapping rules for common (log) sources/formats already exist today in ECS so they can then be used in OTel or are they yet to be defined?

Reviewer: @jkowall @cyrille-leclerc Can you please address the questions above?

Reviewer 2: It makes it much easier to have a schema to normalize the fields and meanings together for correlation. Ultimately, we would want all vendors to log in Otel natively, but that will take a long time. You can do the correlation and parsing in various stages of the pipeline. It might be done in the collector, or as part of an ingestion pipeline to a telemetry service. We see both methods at my organization since we have a sophisticated parsing pipeline. 

There are mapping rules for ECS inside some solutions like Logstash, @cyrille-leclerc maybe you can suggest other implementations of ECS and parsing rules?

Author: Resolving this conversation as @jkowall answered

The coverage of ECS is very broad including in depth support for logs, security, and network events such as "[logs.* fields](https://www.elastic.co/guide/en/ecs/current/ecs-log.html)" , "[geo.* fields](https://www.elastic.co/guide/en/ecs/current/ecs-geo.html)", "[tls.* fields](https://www.elastic.co/guide/en/ecs/current/ecs-tls.html)", "[dns.* fields](https://www.elastic.co/guide/en/ecs/current/ecs-dns.html)", or "[vulnerability.* fields](https://www.elastic.co/guide/en/ecs/current/ecs-vulnerability.html)".

ECS has the following guiding principles:

* ECS favors human readability in order to enable broader adoption,
* ECS events include metadata to enable correlations across any dimension (host, data center, docker image, ip address...),

>     * ECS doesn't differentiate the metadata fields that are specific to each event of the event source and the metadata that are shared by all the events of the source the same way OTEL differentiates Resource Attributes and Log/Span Attributes),

Reviewer: ```suggestion
    * ECS does not differentiate the metadata fields that are specific to each event of the event source and the metadata that is shared by all the events of the source in the way OTEL does, which differentiates between Resource Attributes and Log/Span/Metrics Attributes,
```
To make clear that this is a key difference.

* ECS groups fields in namespaces in order to:
    * Offer consistency and readability,
    * Enable reusability of namespaces in different contexts,

>         * For example, the "geo" namespace is nested in the "client.geo", "destination.geo", "host.geo" or "threat.indicator.geo" namespaces

Reviewer: This is a nice trait that is currently missing in OTel semantic conventions👍

Reviewer 3: I think solving this would also remove one concern I have for nested attributes https://github.com/open-telemetry/opentelemetry-specification/issues/376

Author: Thanks for the feedback @Oberon00 , I'm resolving this comment to move forward

>     * Enable extensibility adding fields to namespaces and adding new namespaces,

Reviewer: ```suggestion
    * Enable extensibility by adding fields to namespaces and adding new namespaces,
```

Prometheus is a de facto standard for observability metrics and OpenTelemetry already provides full interoperability with the Prometheus ecosystem.

It would be useful to get interoperability between metrics collected by [official Prometheus exporters](https://prometheus.io/docs/instrumenting/exporters/) (e.g. the [Node/system metrics exporter](https://github.com/prometheus/node_exporter) or the [MySQL server exporter](https://github.com/prometheus/mysqld_exporter)) and their equivalent OpenTelemetry Collector receivers (e.g. Otel Collector [Host Metrics Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/hostmetricsreceiver) or [MySQL Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/mysqlreceiver)).

Note that one of the challenges with Prometheus metrics naming conventions is that these are implicit conventions defined by each integration author which doesn't enable correlation due to the lack of consistency across integrations. For example, this inconsistency increases the complexity that an end-user has to deal with when configuring and monitoring alerts.

> Prometheus' conventions are restricted to the style of the name of the metrics (see [Metric and label naming | Prometheus](https://prometheus.io/docs/practices/naming/) ) but don't specify unified metric names.

Reviewer: ```suggestion
Prometheus' conventions are restricted to the style of the name of the metrics (see [Prometheus Metric and label naming](https://prometheus.io/docs/practices/naming/)) but don't specify unified metric names.
```
